File: Coding Task Challenge/Question1/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

char caesarRotateChar(char ch, int rot) {
    if (ch >= 'A' && ch <= 'Z')
        return char(( (ch - 'A' + rot) % 26 ) + 'A');
    else if (ch >= 'a' && ch <= 'z')
        return char(( (ch - 'a' + rot) % 26 ) + 'a');
    return ch;
}

string caesarRotateString(const string &s, int rot) {
    string out = "";
    for (char ch : s) out += caesarRotateChar(ch, rot);
    return out;
}

string rleEncode(const string &s) {
    if (s.empty()) return "";
    string out = "";
    char cur = s[0];
    int cnt = 1;

    for (int i = 1; i < s.size(); i++) {
        if (s[i] == cur) cnt++;
        else {
            out += cur;
            if (cnt > 1) out += to_string(cnt);
            cur = s[i];
            cnt = 1;
        }
    }
    // append last group
    out += cur;
    if (cnt > 1) out += to_string(cnt);

    return out;
}

string specialCipher(const string &input, int rot) {
    string rotated = caesarRotateString(input, rot);
    return rleEncode(rotated);
}

int main() {
    cout << specialCipher("AABCCC", 3) << endl; 
    return 0;
}

File: Coding Task Challenge/Question1/Solution.js
// specialCipher.js
function caesarRotateChar(ch, rot) {
  const A = 'A'.charCodeAt(0), Z = 'Z'.charCodeAt(0);
  const a = 'a'.charCodeAt(0), z = 'z'.charCodeAt(0);
  const code = ch.charCodeAt(0);
  if (code >= A && code <= Z) {
    return String.fromCharCode(((code - A + rot) % 26) + A);
  } else if (code >= a && code <= z) {
    return String.fromCharCode(((code - a + rot) % 26) + a);
  }
  return ch;
}

function caesarRotateString(s, rot) {
  return [...s].map(ch => caesarRotateChar(ch, rot)).join('');
}

function rleEncode(s) {
  if (!s) return '';
  let out = '';
  let cur = s[0];
  let cnt = 1;
  for (let i = 1; i < s.length; i++) {
    if (s[i] === cur) cnt++;
    else {
      out += cur + (cnt > 1 ? String(cnt) : '');
      cur = s[i];
      cnt = 1;
    }
  }
  out += cur + (cnt > 1 ? String(cnt) : '');
  return out;
}

function specialCipher(input, rot) {
  const rotated = caesarRotateString(input, rot);
  return rleEncode(rotated);
}

// Example:
console.log(specialCipher("AABCCC", 3)); 
module.exports = { specialCipher };
